### PR TITLE
Plant genetics

### PR DESCRIPTION
**Pull request by Partmedia**
## About the PR

Adds plant genetics. Plants can now be cross-pollinated with other species to produce hybrids, or with each other to transfer desirable traits. Unstable mutagen now causes plants to experience random mutations, and if you're lucky, you'll get a plant with some fun traits.

This PR consists of three commits:

1. Add *Viable*, *Sentient*, *Seedless*, and *Slip* traits. Implements (existing) bioluminescent trait. Adjust the default tolerances to be a bit more reasonable.
2. Implements plant mutation and crossing. Allows you to use mouth swabs to cross-pollinate plants. Adds mouth swabs and hand labelers to the botany locker so botanists can cross-pollinate plants and avoid confusing themselves. Adds a (currently-unused) unstable mutagen prototype for debugging, but suitable for future inclusion in an emagged botany vendor inventory.
3. Nerf robust harvest. To get maximum potency, you must use plant mutations.

**Screenshots**
plant_genetics
*A botanist harvests some bio-luminescent wheat*

**Changelog**
:cl: notafet
- add: Plants can now be mutated using unstable mutagen and cross-pollinated.

### CODE REVIEW

File: Resources/Prototypes/Entities/Objects/Specific/Medical/disease.yml
Comment (Elijahrane) [inline Resources/Prototypes/Entities/Objects/Specific/Medical/disease.yml:17]: having these on the same entity isn't too bad of an idea but you should change name and descrption

File: Resources/Locale/en-US/botany/components/seed-extractor-component.ftl
Comment (Elijahrane) [inline Resources/Locale/en-US/botany/components/seed-extractor-component.ftl:4]: Definite articles like 'the' have a function for them in localization

File: Content.Server/Botany/SeedPrototype.cs
Comment (Elijahrane) [inline Content.Server/Botany/SeedPrototype.cs:129]: What's with all these changes to default tolerances?

File: Content.Server/Botany/Systems/SeedExtractorSystem.cs
Comment (Elijahrane) [inline Content.Server/Botany/Systems/SeedExtractorSystem.cs:34]: By convention a failure should use a caution popup

File: Content.Server/Chemistry/ReagentEffects/PlantMetabolism/RobustHarvest.cs
Comment (Elijahrane) [inline Content.Server/Chemistry/ReagentEffects/PlantMetabolism/RobustHarvest.cs:19]: This should probably be a field on the effect so we can have weaker or stronger reagents with this effect. Same for the hardcoded potency increase

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (Elijahrane) [inline Content.Server/Botany/Systems/MutationSystem.cs:208]: This should probably be a dependency rather than being resolved

**Review by Elijahrane — CHANGES_REQUESTED** [review]

File: Resources/Locale/en-US/botany/components/seed-extractor-component.ftl
Comment (Partmedia) [reply Resources/Locale/en-US/botany/components/seed-extractor-component.ftl:4]: Usage is inconsistent in the existing codebase, so I picked what is done more frequently (explicitly writing out 'the'). A simple search will show that this is the case.

As you may know, there are languages where the definite article (English 'the') must match in number, gender, case, and more with the noun being referred to, e.g. "**the** computer" is "**das** Computer" but (finding something) "by **the** sea" is "an **dem** Meer".

File: Resources/Prototypes/Entities/Objects/Specific/Medical/disease.yml
Comment (Partmedia) [reply Resources/Prototypes/Entities/Objects/Specific/Medical/disease.yml:17]: I'll update this.

File: Content.Server/Botany/SeedPrototype.cs
Comment (Partmedia) [reply Content.Server/Botany/SeedPrototype.cs:129]: I explained this in the PR description: Adjust the default tolerances to be a bit more reasonable.

Basically, if these weren't changed, mutations across these tolerances would not be noticable, i.e. the plant was too tolerant to heat to begin with, so if the heat gene was mutated, you'd see nothing happen.

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (Partmedia) [reply Content.Server/Botany/Systems/MutationSystem.cs:208]: Sure.

File: Resources/Locale/en-US/botany/components/seed-extractor-component.ftl
Comment (Elijahrane) [reply Resources/Locale/en-US/botany/components/seed-extractor-component.ftl:4]: If it is more frequent it's because most of the loc files have not been updated. The definite article function is more recently than you'd think and indefinite articles exist now but are new to the point they are almost unused.

Comment (moonheart08) [reply Resources/Locale/en-US/botany/components/seed-extractor-component.ftl:4]: > Usage is inconsistent in the existing codebase, so I picked what is done more frequently (explicitly writing out 'the'). A simple search will show that this is the case.
> 
> As you may know, there are languages where the definite article (English 'the') must match in number, gender, case, and more with the noun being referred to, e.g. "**the** computer" is "**das** Computer" but (finding something) "by **the** sea" is "an **dem** Meer".

Use the localization function ffs

Comment (Partmedia) [reply Resources/Locale/en-US/botany/components/seed-extractor-component.ftl:4]: Sure, I'll make the change. I just wanted to make sure there were no major structural issues first.

File: Content.Server/Botany/Components/BotanySwabComponent.cs
Comment (mirrorcult) [inline Content.Server/Botany/Components/BotanySwabComponent.cs:9]:
```suggestion
    /// <summary>
    /// Anything that can be used to cross-pollinate plants.
    /// </summary>
    [RegisterComponent]
    public sealed class BotanySwabComponent : Component
```
convention

File: Content.Server/Botany/SeedPrototype.cs
Comment (mirrorcult) [inline Content.Server/Botany/SeedPrototype.cs:177]: not sure how i feel about hardcoding all of these possible variations into the seedprototype. eventually there's going to be a lot more of these, and they're going to want some variation (radius, color, degree of slipperiness...)

a solution I think we could do is something similar to `TraitPrototype` which just defines a component registry and adds everything onto the entity that spawns. so the seedprototype could just have a list of `BotanyTraitPrototype` IDs and it adds all the components from those on. that way you can have a lot more variation easily

File: Content.Server/Botany/Components/PlantHolderComponent.cs
Comment (mirrorcult) [inline Content.Server/Botany/Components/PlantHolderComponent.cs:562]: invert the if in that case then

File: Content.Server/Botany/Components/PlantHolderComponent.cs
Comment (mirrorcult) [inline Content.Server/Botany/Components/PlantHolderComponent.cs:394]: so based

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (mirrorcult) [inline Content.Server/Botany/Systems/MutationSystem.cs:53]: i dont think mysterious is meant to be altered

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (mirrorcult) [inline Content.Server/Botany/Systems/MutationSystem.cs:190]: why not just truly randomize the color? was this tested and it just looked bad aesthetically (i know some color systems fare badly with pure randomization)

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (mirrorcult) [inline Content.Server/Botany/Systems/MutationSystem.cs:52]: any reason why `MutateBool` only goes for one side (i.e. why these can't be flipped off with a mutation?)

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (mirrorcult) [inline Content.Server/Botany/Systems/MutationSystem.cs:103]: is it a thermometer code if you're 'flipping' them? i thought thermometer codes were unary?

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (mirrorcult) [inline Content.Server/Botany/Systems/MutationSystem.cs:132]: does `MathF.Min` et al allow you to remove the float casts

**Review by mirrorcult — CHANGES_REQUESTED** [review]
on overall design:
- sterile swab and crossing is really neat, and im glad that it uses the same medical swabs
- mutation approach is really really cool, just have some issues with it being potentially hardcodey

implementation concept looks very very good

File: Content.Server/Botany/SeedPrototype.cs
Comment (Partmedia) [reply Content.Server/Botany/SeedPrototype.cs:177]: While this could be desirable in the future when there are more traits, hard coding this now separates the content changes from refactoring changes and reduces the diff of this PR.

While I'd be happy to work on a `TraitPrototype`, perhaps it could be better organized into a future PR.

File: Content.Server/Botany/Components/PlantHolderComponent.cs
Comment (Partmedia) [reply Content.Server/Botany/Components/PlantHolderComponent.cs:562]: I'll make this change.

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (Partmedia) [reply Content.Server/Botany/Systems/MutationSystem.cs:53]: I'll remove mysterious alteration.

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (Partmedia) [reply Content.Server/Botany/Systems/MutationSystem.cs:190]: These colors are interesting and distinguishable. With the small light radius, it can be hard to discern certain color glows (e.g. rolling grey would just show up as a dimmer white). As you point out, what I probably want here is randomizing only 'H' in the HSV color space, and converting back to RGB.

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (Partmedia) [reply Content.Server/Botany/Systems/MutationSystem.cs:52]: To make it more likely that you get an observable boolean mutation. If there was one bit with an equal chance of flipping back and forth, it would tend towards a coin toss and be less interesting. With this bias, you are guaranteed to get more mutations as long as you add more unstable mutagen.

Seedless and Viable traits are intended to "nerf" mutation: seedless by making mutated plants not have seeds (you can't just clone them), and Viable to kill the plants outright. Having those particular traits mutate back would un-nerf the nerf.

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (Partmedia) [reply Content.Server/Botany/Systems/MutationSystem.cs:103]: The underlying idea is that if you have a trait representable by '11100' (a true thermometer code), flipping one bit gives, as an example, '11101' (not a thermometer code), which you can sort and get '1110' (a thermometer code again).

Statistically, this gives an expected final distribution that trends to the middle of where the thermometer code is. And it is also a thermometer code in the sense that the minimum change is one least-significant bit. But it might just be poor naming.

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (Partmedia) [reply Content.Server/Botany/Systems/MutationSystem.cs:132]: Probably. I didn't know that MathF exists. I'll make the change.

File: Content.Server/Botany/Components/PlantHolderComponent.cs
Comment (Partmedia) [reply Content.Server/Botany/Components/PlantHolderComponent.cs:394]: Thank you for the careful review and thoughtful feedback. I'll update the PR to address the issues I responded to above and wait for feedback on the others.

**Comment by Partmedia** [comment]
Thank you for the careful review and thoughtful feedback. I'll update the PR to address the issues I responded to above and wait for feedback on the others.

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (mirrorcult) [reply Content.Server/Botany/Systems/MutationSystem.cs:52]: okay, that makes sense

File: Content.Server/Botany/SeedPrototype.cs
Comment (mirrorcult) [reply Content.Server/Botany/SeedPrototype.cs:177]: alright, i'm fine with that coming later

File: Content.Server/Botany/Systems/MutationSystem.cs
Comment (Partmedia) [reply Content.Server/Botany/Systems/MutationSystem.cs]: The outer float cast goes away. The inner float cast has to stay to coerce 'np' from integer to float.

**Comment by Partmedia** [comment]
So, I neglected to test plant crossing after implementing one of the reviewer's feedback. Specifically, using dependencies (i.e. [dependency] _robustRandom) instead of IoCManager causes an unhandled NullReferenceException that isn't found by the nullable checker:

```
[FATL] unhandled: System.NullReferenceException: Object reference not set to an instance of an object.
   at Robust.Shared.Random.RandomExtensions.Prob(IRobustRandom random, Single chance) in /space-station-14/RobustToolbox/Robust.Shared/Random/RandomExtensions.cs:line 81
   at Content.Server.Botany.MutationSystem.random(Single p) in /space-station-14/Content.Server/Botany/Systems/MutationSystem.cs:line 230
   at Content.Server.Botany.MutationSystem.Cross(SeedData a, SeedData b) in /space-station-14/Content.Server/Botany/Systems/MutationSystem.cs:line 81
   at Content.Server.Botany.BotanySwabSystem.OnTargetSwabSuccessful(TargetSwabSuccessfulEvent args) in /space-station-14/Content.Server/Botany/Systems/BotanySwabSystem.cs:line 97
```

Even worse, this bit gets optimized away in the release build, so from the backtrace on a running production server it's not even clear what's going on.

I'm not sure why this happens, but I suspect that it has something to do with the fact that MutationSystem is not a real EntitySystem, but rather a class with static helper functions. Someone please correct me if I'm wrong, or explain why this happens if I'm right.

I've reverted the random change for now, and the code works as expected.

A reminder that bugs can creep in *during* the review process, pointyhat to me for not checking...

**Comment by mirrorcult** [comment]
> I'm not sure why this happens, but I suspect that it has something to do with the fact that MutationSystem is not a real EntitySystem, but rather a class with static helper functions. Someone please correct me if I'm wrong, or explain why this happens if I'm right.
> 

yes, IoC dependencies are automatically injected into entitysystems, but not other classes without `IoCManager.InjectDependencies(this)` being called first

**Comment by Partmedia** [comment]
> > I'm not sure why this happens, but I suspect that it has something to do with the fact that MutationSystem is not a real EntitySystem, but rather a class with static helper functions. Someone please correct me if I'm wrong, or explain why this happens if I'm right.
> 
> yes, IoC dependencies are automatically injected into entitysystems, but not necessarily other classes (there are some other special ones) without `IoCManager.InjectDependencies(this)` being called first

Thanks for the explanation. Since MutationSystem is not an EntitySystem, there doesn't appear to be a good place (like an Init() function) for such a InjectDependencies call. Should I just keep grabbing random directly from IoC?

**Comment by mirrorcult** [comment]
you could always just make it an entitysystem and de-static the methods

**Comment by Partmedia** [comment]
> you could always just make it an entitysystem and de-static the methods

Your call. I put these functions here because Cross() and Mutate() felt like they belonged with each other, instead of one in the BotanySwabSystem and the other in PlantHolderComponent. The botany code also looked like it was about to get a ECS refactor, so I wanted to put the mutation code somewhere it wouldn't be affected.

**Comment by mirrorcult** [comment]
yeah, i think making `MutationSystem` an actual system is the best call here, though it means you need to resolve it in other places

**Comment by Partmedia** [comment]
`MutationSystem` converted to an actual entity system.

File: Content.Server/IoC/ServerContentIoC.cs
Comment (mirrorcult) [inline Content.Server/IoC/ServerContentIoC.cs:61]: this is not necessary, entity systems are registered normally

Comment (mirrorcult) [reply Content.Server/IoC/ServerContentIoC.cs:61]: i imagine this was done to have it be a dep in `PlantHolderComponent`--in that case, just call `_entMan.System<MutationSystem>` where necessary instead

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Content.Server/IoC/ServerContentIoC.cs
Comment (Partmedia) [reply Content.Server/IoC/ServerContentIoC.cs]: Fixed, thanks.

File: Content.Server/Botany/Systems/BotanySwabSystem.cs
Comment (mirrorcult) [inline Content.Server/Botany/Systems/BotanySwabSystem.cs:22]: system dependencies are fine in other systems, just in the component was bad. this can be changed back, sorry if i was unclear

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Content.Server/Botany/Systems/BotanySwabSystem.cs
Comment (Partmedia) [reply Content.Server/Botany/Systems/BotanySwabSystem.cs]: Fixed, thanks.

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**
